### PR TITLE
chore: Move environment config to TS

### PR DESCRIPTION
**Pull request by ds300**
This PR resolves [CX-871]

### Description

This PR moves the app's [environment configuration](https://github.com/artsy/eigen/pull/4465/files#diff-13f9506e86aecda94da8d225c6be033bf731681310428cee9b039ee612c10b15) to TS.

This was quite a significant change that had a lot of knock-on effects and related fixes:

- Consolidated sentry projects into one (previously 'eigen-beta' and 'eigen'. Now just 'eigen'. See [slack context](https://artsy.slack.com/archives/C02BAQ5K7/p1613498325159700))
- Moved [auth validation check](https://github.com/artsy/eigen/pull/4465/files#diff-62a4f62f32cb84f82db8a186424bd7cce489a9575d2888db316cb90fdf25e995) to TS. (this check is responsible for logging the user out with a friendly Alert if the user's session expires due to password changes etc)
- Refactored the `ConfigModel` into separate models for feature flags, echo state management, and env state management.
- Simplified sign out control flow.
  - It used to be a cyclical thing where we'd dispatch an event requesting a logout through ARNotificationsManager, then the native user state would be cleared and it would dispatch an event back over the bridge to clear the easy peasy state. 
  - Now [we clear the easy peasy state directly](https://github.com/artsy/eigen/blob/902a8c74c8d0ac0ef0f3680aa084cd9d6c92cfb2/src/lib/store/GlobalStoreModel.ts#L54-L63) (without clearing admin settings) at the same time as asking `ARUserManager` to clear its state
- Added a `TS -> objc` [state bridge for env strings](https://github.com/artsy/eigen/blob/902a8c74c8d0ac0ef0f3680aa084cd9d6c92cfb2/emission/Pod/Classes/EigenCommunications/ARNotificationsManager.m#L25-L32). (now the only obj-c controlled state is auth/onboarding-related, so we'll be fully TS when the new onboarding stuff is deployed on iOS)
- Added basic admin menu functionality for toggling between staging and production, and for using local URLs in dev mode.

https://user-images.githubusercontent.com/1242537/108872197-d5d4e700-75f1-11eb-82d1-9666d3f2044b.mp4

https://user-images.githubusercontent.com/1242537/108873055-ba1e1080-75f2-11eb-9671-f0635b631f9d.mp4

Things to QA (Completed 💆‍♀️)

- [x] Sign out (iOS and android)
- [x] Switch between staging and production (iOS and android)
- [x] Inquiry (iOS only)
- [x] Sentry error reporting (add env tag)
- [x] Share artwork (copy link)  (iOS only)
- [x] FairEditorialShare links (iOS only)
- [x] BNMO web view (iOS only)
- [x] Feature flags & echo (iOS and android)
- [x] Add new credit card (iOS only)

### Follow-up work

- Find a way to hide the admin menu in releases [Jira](https://artsyproduct.atlassian.net/browse/CX-176)
- Move the launcher section from the old admin menu to the new admin menu. [Jira](https://artsyproduct.atlassian.net/browse/CX-1112)

### PR Checklist (tick all before merging)

- [x] I have included screenshots or videos to illustrate my changes, or I have not changed anything that impacts the UI.
- [x] I have added tests for my changes, or my changes don't require testing, or I have included a link to a separate Jira ticket covering the tests.
- [x] I have documented any follow-up work that this PR will require, or it does not require any.
- [x] I have added an app state migration, or my changes do not require one. ([What are migrations?](https://github.com/artsy/eigen/blob/master/docs/adding_state_migrations.md))
- [x] I have added a [CHANGELOG.yml](/CHANGELOG.yml) entry or my changes do not require one.

[CX-871]: https://artsyproduct.atlassian.net/browse/CX-871

### CODE REVIEW

File: src/lib/NativeModules/LegacyNativeModules.tsx [hunk 1]
```diff
@@ -72,46 +80,49 @@ interface LegacyNativeModules {
...
+export const LegacyNativeModules: LegacyNativeModules =
+  Platform.OS === "ios"
+    ? LegacyNativeModulesIOS
+    : {
+        ARTakeCameraPhotoModule: {
+          errorCodes: {
+            cameraNotAvailable: "cameraNotAvailable",
+            imageMediaNotAvailable: "imageMediaNotAvailable",
+            cameraAccessDenied: "cameraAccessDenied",
+            saveFailed: "saveFailed",
```
Comment (ds300) [inline src/lib/NativeModules/LegacyNativeModules.tsx:94]: apologies for the big diff here. I changed the way this const is declared because the type checking wasn't working with `Platform.select`

File: src/lib/relay/createEnvironment.ts [hunk 2]
```diff
@@ -38,9 +38,6 @@ export default function createEnvironment() {
       // session is still valid. This is why we need to keep it as low as possible in the middlewares array.
       checkAuthenticationMiddleware(),
       loggerMiddleware(),
-      errorMiddleware({
-        disableServerMiddlewareTip: true,
-      }),
```
Comment (ds300) [inline src/lib/relay/createEnvironment.ts:43]: removed this useless ancient error logging middleware which breaks with Fast Refresh (Ends up logging errors that happened a while ago on every subsequent request). Relay logs errors perfectly well on it's own these days.

File: src/lib/store/AuthModel.ts [hunk 3]
```diff
@@ -124,13 +122,4 @@ export const AuthModel: AuthModel = {
...
   }),
-  signOut: action((state) => {
-    Object.assign(state, {
-      userID: null,
-      userAccessToken: null,
-      userAccessTokenExpiresIn: null,
-      xAppToken: null,
-      xApptokenExpiresIn: null,
-    })
-  }),
```
Comment (ds300) [inline src/lib/store/AuthModel.ts:135]: `signOut` moved to the main `GlobalStoreModel`.

File: src/lib/store/__tests__/ConfigModel-tests.ts [hunk 4]
```diff
@@ -1,139 +0,0 @@
-import fetchMock from "jest-fetch-mock"
```
Comment (ds300) [inline src/lib/store/__tests__/ConfigModel-tests.ts:1]: the tests in this file were moved into other files

File: src/lib/utils/volleyClient.ts [hunk 5]
```diff
@@ -66,6 +65,9 @@ export type VolleyMetric =
 
 class VolleyClient {
   queue: VolleyMetric[] = []
+  get serviceName() {
+    return unsafe__getEnvironment().env === "staging" ? "eigen-staging" : "eigen"
+  }
```
Comment (ds300) [inline src/lib/utils/volleyClient.ts:70]: this used to be passed in as a parameter to the VolleyClient constructor, which is a singleton. So now that the env can change at run time we need to check it on-demand.

File: Artsy/App/ARAppDelegate.m [hunk 6]
```diff
@@ -386,13 +384,6 @@ - (void)rageShakeNotificationRecieved
 - (void)countNumberOfRuns
 {
     NSInteger numberOfRuns = [[NSUserDefaults standardUserDefaults] integerForKey:ARAnalyticsAppUsageCountProperty] + 1;
-    if (numberOfRuns == 1) {
-        [ARAnalytics event:ARAnalyticsFreshInstall];
-    }
-
-    if (numberOfRuns == 3) {
-        [[ARUserManager sharedManager] tryStoreSavedCredentialsToWebKeychain];
-    }
```
Comment (ds300) [inline Artsy/App/ARAppDelegate.m:395]: these moved to ARAppDelegate+Emission because they need to wait until after emission has bootstrapped to use ARRouter.

File: Artsy/App/ARAppDelegate+Emission.m [hunk 7]
```diff
@@ -47,15 +46,16 @@ @implementation ARAppDelegate (Emission)
...
+    isDebugMode = YES;
+#else
+    isDebugMode = NO;
+#endif
+    if (isDebugMode) {
         NSString *bundleUrlString = [NSString stringWithFormat:@"http://%@:8081/index.ios.bundle?platform=ios&dev=true", [ARReactPackagerHost hostname]];
         NSURL *packagerURL = [NSURL URLWithString:bundleUrlString];
         [self setupSharedEmissionWithPackagerURL:packagerURL];
-
     } else {
```
Comment (ds300) [inline Artsy/App/ARAppDelegate+Emission.m:59]: Before the emission + eigen combination, we supported loading the JS bundle from different places. Now this will behave like a normal react native app, i.e.

- at dev time, use the packager server
- otherwise use the js bundled with the app

File: src/lib/AndroidApp.tsx [hunk 8]
```diff
@@ -5,18 +5,23 @@ import { Linking, View } from "react-native"
...
 import { AdminMenuWrapper } from "./utils/AdminMenuWrapper"
 import { ProvideScreenDimensions } from "./utils/useScreenDimensions"
+import { useStripeConfig } from "./utils/useStripeConfig"
 
 const Main: React.FC<{}> = track()(({}) => {
   const isHydrated = GlobalStore.useAppState((state) => state.sessionState.isHydrated)
   const isLoggedIn = GlobalStore.useAppState((state) => !!state.auth.userAccessToken)
 
+  useSentryConfig()
+  useStripeConfig()
```
Comment (ds300) [inline src/lib/AndroidApp.tsx:23]: Couple of new hooks to configure sentry and stripe based on changes to the environment (mainly switching between staging and prod), but also the userID for sentry.

File: src/lib/Scenes/BottomTabs/__tests__/BottomTabsNavigator-tests.tsx [hunk 9]
```diff
@@ -14,7 +15,9 @@ describe(BottomTabsNavigator, () => {
       tree.root.findAll((node) => node.type === NativeViewController && node.props.viewProps.tabName === "search")
     ).toHaveLength(0)
 
-    GlobalStore.actions.bottomTabs.switchTab("search")
+    await act(() => {
+      GlobalStore.actions.bottomTabs.switchTab("search")
+    })
```
Comment (ds300) [inline src/lib/Scenes/BottomTabs/__tests__/BottomTabsNavigator-tests.tsx:20]: this change is unrelated, I just noticed that react was logging a warning about not using act here

File: src/lib/store/GlobalStore.tsx [hunk 10]
```diff
@@ -164,9 +158,26 @@ export function unsafe__getSelectedTab() {
 }
 
 export function useIsStaging() {
-  // TODO: support non-staging environments on android
```
Comment (ds300) [inline src/lib/store/GlobalStore.tsx:167]: ✔️

File: src/lib/store/GlobalStoreModel.ts [hunk 11]
```diff
@@ -1,18 +1,22 @@
...
 import { CURRENT_APP_VERSION } from "./migration"
 import { NativeModel } from "./NativeModel"
-import { assignDeep } from "./persistence"
+import { assignDeep, sanitize } from "./persistence"
 
 interface GlobalStoreStateModel {
   version: number
   sessionState: {
     isHydrated: boolean
+    didSessionExpire: boolean
```
Comment (MounirDhahri) [inline src/lib/store/GlobalStoreModel.ts:19]: I think you haven't used this anywhere

**Review by MounirDhahri — APPROVED** [review]

**Comment by MounirDhahri** [comment]
Good job! What a nice PR

File: src/lib/utils/AdminMenu.tsx [hunk 12]
```diff
@@ -158,3 +174,99 @@ const FeatureFlagItem: React.FC<{ flagKey: FeatureName }> = ({ flagKey }) => {
...
+  return {
+    text,
+    onPress() {
+      GlobalStore.actions.config.environment.clearAdminOverrides()
+      if (env !== currentEnv) {
+        GlobalStore.actions.config.environment.setEnv(env)
+        if (Platform.OS === "ios") {
+          dismissModal()
+          LegacyNativeModules.ARNotificationsManager.postNotificationName("ARUserRequestedLogout", {})
+        } else {
```
[the remark this answers is not in the document]
Comment (ds300) [reply src/lib/utils/AdminMenu.tsx]: i think you're looking at an early commit there, check the latest revision: https://github.com/artsy/eigen/pull/4465/files

File: src/lib/store/GlobalStoreModel.ts (hunk 11 above)
Comment (ds300) [reply src/lib/store/GlobalStoreModel.ts]: good catch! 🙌  removed

File: Artsy/App/ARAppDelegate+Emission.m [hunk 13]
```diff
@@ -140,19 +108,22 @@ - (void)setupSharedEmissionWithPackagerURL:(NSURL *)packagerURL;
...
+        [ARRouter setup];
+        
+        if (launchCount == 1) {
+            [ARAnalytics event:ARAnalyticsFreshInstall];
+        }
+
+        if (launchCount == 3) {
+            [[ARUserManager sharedManager] tryStoreSavedCredentialsToWebKeychain];
+        }
+    }];
```
[the remark this answers is not in the document]
Comment (ds300) [reply Artsy/App/ARAppDelegate+Emission.m:124]: ah yes, thanks for adding that context!

File: .env.example [hunk 14]
```diff
@@ -9,8 +9,7 @@ ARTSY_API_CLIENT_KEY=e750db60ac506978fc70
 ARTSY_FACEBOOK_APP_ID=-
 SEGMENT_PRODUCTION_WRITE_KEY=-
 SEGMENT_STAGING_WRITE_KEY=-
-SENTRY_PRODUCTION_DSN=-
-SENTRY_BETA_DSN=-
+SENTRY_DSN=-
```
[the remark this answers is not in the document]
Comment (ds300) [reply .env.example:12]: ooh good catch, I had completely forgotten about that file

[the remark this answers is not in the document]
Comment (ds300) [reply .env.example:12]: ok I updated the file

File: src/lib/utils/AdminMenu.tsx (hunk 12 above)
[the remark this answers is not in the document]
Comment (MounirDhahri) [reply src/lib/utils/AdminMenu.tsx]: oh yes, thanks a lot

File: src/lib/store/ConfigModel.ts [hunk 15]
```diff
@@ -1,88 +1,14 @@
...
-      }
-      actions.fetchRemoteEcho()
-    }
-  ),
-  setAdminOverride: action((state, { key, value }) => {
-    state.adminFeatureFlagOverrides[key] = value
-  }),
+  echo: EchoModel,
+  features: FeaturesModel,
+  environment: EnvironmentModel,
```
Comment (brainbicycle) [inline src/lib/store/ConfigModel.ts:13]: ✨

File: src/lib/store/config/FeaturesModel.tsx [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+      } else if (feature.readyForRelease) {
+        // If the feature is ready for release, the echo flag takes precedence
+        const echoFlag = echo.json.features.find((f) => f.name === feature.echoFlagKey)
+
+        if (feature.echoFlagKey && !echoFlag && __DEV__) {
+          console.error("No echo flag found for feature", key)
+        }
+        result[key] = echoFlag?.value ?? true
+      } else {
+        // If the feature is not ready for release, uh, don't show it
```
Comment (brainbicycle) [inline src/lib/store/config/FeaturesModel.tsx:37]: 😆

File: src/lib/store/config/__tests__/FeaturesModel-tests.tsx [hunk 17]
```diff
@@ -0,0 +1,86 @@
...
+    GlobalStore.actions.config.echo.setEchoState({
+      ...echoLaunchJSON,
+      features: [
+        { name: "FeatureAEchoKey", value: false },
+        { name: "FeatureBEchoKey", value: true },
+      ],
+    })
+    expect(getComputedFeatures().FeatureB).toBe(false)
+  })
+})
```
Comment (brainbicycle) [inline src/lib/store/config/__tests__/FeaturesModel-tests.tsx:87]: 🎉

File: Artsy/App/ARAppDelegate+Emission.m (hunk 13 above)
[the remark this answers is not in the document]
Comment (brainbicycle) [reply Artsy/App/ARAppDelegate+Emission.m:124]: Nice find!

File: src/lib/Scenes/VanityURL/__tests__/VanityURLPossibleRedirect-tests.tsx [hunk 18]
```diff
@@ -12,7 +12,7 @@ import { VanityURLPossibleRedirect } from "../VanityURLPossibleRedirect"
 const fetchMock = ((global as any).fetch = jest.fn())
 
 beforeEach(() => {
-  __globalStoreTestUtils__?.injectState({ native: { sessionState: { webURL: "https://artsy.net" } } })
+  __globalStoreTestUtils__?.setProductionMode()
```
Comment (pvinis) [inline src/lib/Scenes/VanityURL/__tests__/VanityURLPossibleRedirect-tests.tsx:15]: squeeky clean 🙆‍♂️

File: Artsy/Networking/ARRouter.m [hunk 19]
```diff
@@ -44,10 +46,9 @@ + (void)setup
     NSString *productionDeprecatedMobileWeb = hostFromString(ARBaseDeprecatedMobileWebURL);
     NSString *productionAPI = hostFromString(ARBaseApiURL);
 
-    NSUserDefaults *defaults = [NSUserDefaults standardUserDefaults];
-    NSString *stagingWeb = hostFromString([defaults stringForKey:ARStagingWebURLDefault]);
+    NSString *stagingWeb = @"staging.artsy.net";
     NSString *stagingDeprecatedMobileWeb = hostFromString(ARStagingBaseDeprecatedMobileWebURL);
-    NSString *stagingAPI = hostFromString([defaults stringForKey:ARStagingAPIURLDefault]);
```
Comment (brainbicycle) [inline Artsy/Networking/ARRouter.m:50]: **nitpick (non-blocking): I think this is only place we were using hostFromString method, can we delete it if so?**

File: src/lib/AppRegistry.tsx [hunk 20]
```diff
@@ -403,7 +403,7 @@ const Main: React.FC<{}> = track()(({}) => {
   const onboardingState = GlobalStore.useAppState((state) => state.native.sessionState.onboardingState)
 
   if (!isHydrated) {
-    return <View />
+    return <View style={{ backgroundColor: "papayawhip" }} />
```
Comment (brainbicycle) [inline src/lib/AppRegistry.tsx]: 🇲🇽

**Review by brainbicycle — APPROVED** [review]
Looks great! If we are consolidating sentry projects as part of this PR I think we will also need to update the fastfile to upload sourcemaps and release artifacts to just the eigen sentry project rather than the `eigen-beta` and `eigen` projects, that happens here:
https://github.com/artsy/eigen/blob/655382a277f0c959c1d5603bd29c3391229f6f53/fastlane/Fastfile#L145

File: src/lib/store/config/EchoModel.tsx [hunk 21]
```diff
@@ -0,0 +1,53 @@
...
+  }),
+  fetchRemoteEcho: thunk(async (actions) => {
+    const result = await fetch("https://echo.artsy.net/Echo.json")
+    if (result.ok) {
+      const json = await result.json()
+      actions.setEchoState(json)
+    }
+  }),
+  didRehydrate: thunkOn(
+    (_, storeActions) => storeActions.rehydrate,
```
Comment (damassi) [inline src/lib/store/config/EchoModel.tsx:31]: its useful 😄

**Comment by damassi** [comment]
💯

File: Artsy/Networking/ARRouter.m (hunk 19 above)
Comment (ds300) [reply Artsy/Networking/ARRouter.m:50]: It is still used 4 times in this file. We could hard-code the host strings I suppose, but I'm fine either way so will pick the path of least effort and leave it like this 🐢

Comment (brainbicycle) [reply Artsy/Networking/ARRouter.m:50]: ahh sorry, never mind
